cpp/src/arrow/array.h

```diff
@@ -58,6 +59,9 @@ class Array {
 
   bool EqualsExact(const Array& arr) const;
   virtual bool Equals(const std::shared_ptr<Array>& arr) const = 0;
+  // Determines if the array is internally consistent.  Defaults to always
+  // returning Status::OK.  This can be an expensive check.
+  virtual Status Validate() const;
```

Comment:
There was some work last year in Impala in which they analyzed the generated x86 instructions to see how to minimize the CPU cycles associated with verifying Status::OK: https://github.com/apache/incubator-impala/commit/1afe72830a696e700ac5948ac6e2e0994c50f991#diff-ca5b27977d768460882519dbb9162a4e